### PR TITLE
Add Alpine variant

### PR DESCRIPTION
**Pull request by J0WI**
Closes #49

### CODE REVIEW

**Comment by J0WI** [comment]
I found the issue: https://github.com/naitoh/rbpdf/issues/31
The precompiled binary is not compatible with the musl libc based Alpine Linux. I added `ttf2ufm` manually for now but I'm not sure if it's even used. However I successfully exported wiki articles as PDF with this.

@tianon would be great if you could review this.

**Comment by J0WI** [comment]
Friendly reminder that this is ready to review :smiley: @yosifkit @tianon

**Comment by tianon** [comment]
```diff
$ diff -u Dockerfile-debian.template Dockerfile-alpine.template
--- Dockerfile-debian.template	2019-07-23 14:01:26.549472146 -0700
+++ Dockerfile-alpine.template	2019-07-23 14:02:13.276176948 -0700
@@ -1,12 +1,15 @@
-FROM ruby:%%RUBY_VERSION%%-slim-buster
+FROM ruby:%%RUBY_VERSION%%-alpine3.10
 
 # add our user and group first to make sure their IDs get assigned consistently, regardless of whatever dependencies get added
-RUN groupadd -r redmine && useradd -r -g redmine redmine
+RUN addgroup -S redmine && adduser -S -H -G redmine redmine
 
 RUN set -eux; \
-	apt-get update; \
-	apt-get install -y --no-install-recommends \
+	apk add --no-cache \
+		bash \
 		ca-certificates \
+		su-exec \
+		tini \
+		tzdata \
 		wget \
 		\
 		bzr \
@@ -16,53 +19,10 @@
 		subversion \
 		\
 # https://github.com/docker-library/redmine/issues/132
-# (without "gsfonts" we get "Magick::ImageMagickError (non-conforming drawing primitive definition `text' @ error/draw.c/DrawImage/3265):")
-		gsfonts \
-		imagemagick \
-	; \
-	rm -rf /var/lib/apt/lists/*
-
-RUN set -eux; \
-	savedAptMark="$(apt-mark showmanual)"; \
-	apt-get update; \
-	apt-get install -y --no-install-recommends \
-		dirmngr \
-		gnupg \
-	; \
-	rm -rf /var/lib/apt/lists/*; \
-	\
-	dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')"; \
-	\
-# grab gosu for easy step-down from root
-# https://github.com/tianon/gosu/releases
-	export GOSU_VERSION='1.11'; \
-	wget -O /usr/local/bin/gosu "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch"; \
-	wget -O /usr/local/bin/gosu.asc "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch.asc"; \
-	export GNUPGHOME="$(mktemp -d)"; \
-	gpg --batch --keyserver hkps://keys.openpgp.org --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4; \
-	gpg --batch --verify /usr/local/bin/gosu.asc /usr/local/bin/gosu; \
-	gpgconf --kill all; \
-	rm -r "$GNUPGHOME" /usr/local/bin/gosu.asc; \
-	chmod +x /usr/local/bin/gosu; \
-	gosu nobody true; \
-	\
-# grab tini for signal processing and zombie killing
-# https://github.com/krallin/tini/releases
-	export TINI_VERSION='0.18.0'; \
-	wget -O /usr/local/bin/tini "https://github.com/krallin/tini/releases/download/v$TINI_VERSION/tini-$dpkgArch"; \
-	wget -O /usr/local/bin/tini.asc "https://github.com/krallin/tini/releases/download/v$TINI_VERSION/tini-$dpkgArch.asc"; \
-	export GNUPGHOME="$(mktemp -d)"; \
-	gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys 6380DC428747F6C393FEACA59A84159D7001A4E5; \
-	gpg --batch --verify /usr/local/bin/tini.asc /usr/local/bin/tini; \
-	gpgconf --kill all; \
-	rm -r "$GNUPGHOME" /usr/local/bin/tini.asc; \
-	chmod +x /usr/local/bin/tini; \
-	tini -h; \
-	\
-# reset apt-mark's "manual" list so that "purge --auto-remove" will remove all build dependencies
-	apt-mark auto '.*' > /dev/null; \
-	[ -z "$savedAptMark" ] || apt-mark manual $savedAptMark; \
-	apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false
+# (without "ghostscript-fonts" we get "Magick::ImageMagickError (non-conforming drawing primitive definition `text' @ error/draw.c/DrawImage/3265):")
+		ghostscript-fonts \
+		imagemagick6 \
+	;
 
 ENV RAILS_ENV production
 WORKDIR /usr/src/redmine
@@ -94,46 +54,45 @@
 
 RUN set -eux; \
 	\
-	savedAptMark="$(apt-mark showmanual)"; \
-	apt-get update; \
-	apt-get install -y --no-install-recommends \
+	apk add --no-cache --virtual .build-deps \
+		coreutils \
 		freetds-dev \
 		gcc \
-		libmagickcore-dev \
-		libmagickwand-dev \
-		libmariadbclient-dev \
-		libpq-dev \
-		libsqlite3-dev \
+		imagemagick6-dev \
 		make \
+		mariadb-dev \
+		musl-dev \
 		patch \
+		postgresql-dev \
+		sqlite-dev \
+		ttf2ufm \
+		zlib-dev \
 	; \
-	rm -rf /var/lib/apt/lists/*; \
 	\
-	gosu redmine bundle install --without development test; \
+	su-exec redmine bundle install --without development test; \
 	for adapter in mysql2 postgresql sqlserver sqlite3; do \
 		echo "$RAILS_ENV:" > ./config/database.yml; \
 		echo "  adapter: $adapter" >> ./config/database.yml; \
-		gosu redmine bundle install --without development test; \
+		su-exec redmine bundle install --without development test; \
 		cp Gemfile.lock "Gemfile.lock.${adapter}"; \
 	done; \
 	rm ./config/database.yml; \
 # fix permissions for running as an arbitrary user
 	chmod -R ugo=rwX Gemfile.lock "$GEM_HOME"; \
+# this requires coreutils because "chmod +X" in busybox will remove +x on files (and coreutils leaves files alone with +X)
 	rm -rf ~redmine/.bundle; \
 	\
-# reset apt-mark's "manual" list so that "purge --auto-remove" will remove all build dependencies
-	apt-mark auto '.*' > /dev/null; \
-	[ -z "$savedAptMark" ] || apt-mark manual $savedAptMark; \
-	find /usr/local -type f -executable -exec ldd '{}' ';' \
-		| awk '/=>/ { print $(NF-1) }' \
-		| sort -u \
-		| grep -v '^/usr/local/' \
-		| xargs -r dpkg-query --search \
-		| cut -d: -f1 \
+# https://github.com/naitoh/rbpdf/issues/31
+	rm /usr/local/bundle/gems/rbpdf-font-1.19.*/lib/fonts/ttf2ufm/ttf2ufm; \
+	\
+	runDeps="$( \
+		scanelf --needed --nobanner --format '%n#p' --recursive /usr/local/bundle/gems \
+		| tr ',' '\n' \
 		| sort -u \
-		| xargs -r apt-mark manual \
-	; \
-	apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false
+		| awk 'system("[ -e /usr/local/lib/" $1 " ]") == 0 { next } { print "so:" $1 }' \
+	)"; \
+	apk add --virtual .redmine-rundeps $runDeps; \
+	apk del .build-deps
 
 VOLUME /usr/src/redmine/files
 
```

File: Dockerfile-alpine.template [hunk 1]
```diff
@@ -0,0 +1,103 @@
+FROM ruby:%%RUBY_VERSION%%-alpine3.10
+
+# add our user and group first to make sure their IDs get assigned consistently, regardless of whatever dependencies get added
+RUN addgroup -S redmine && adduser -S -H -G redmine redmine
```
Comment (yosifkit) [inline Dockerfile-alpine.template]: Should we specify user and group IDs to ensure easy moving from one os variant to another? (i.e. use whatever the current IDs debian has and make them static for both)

Comment (yosifkit) [reply Dockerfile-alpine.template]: Unfortunately we can only line up the user id. :cry:
```console
$ docker run --rm --user redmine redmine id
uid=999(redmine) gid=999(redmine) groups=999(redmine)
$ docker run --rm alpine:3.10 sh -c ' cat /etc/passwd | grep 999'
$ docker run --rm alpine:3.10 sh -c ' cat /etc/group | grep 999'
ping:x:999:
```

And add redmine to the `ping` group? :confused: :nauseated_face:

**Comment by yosifkit** [comment]
Added static uid/gid to ensure easier transition for users moving between debian and alpine.

**Comment by J0WI** [comment]
Thanks! :tada:

**Pull request merged**